design/sketch.webidl

```diff
@@ -0,0 +1,532 @@
+
+    void draw(u32 vertexCount, u32 instanceCount, u32 firstVertex, u32 firstInstance);
+    void drawIndexed(u32 indexCount, u32 instanceCount, u32 firstIndex, u32 firstInstance, u32 firstVertex);
+
+    // TODO add missing commands
+};
+
+// ****************************************************************************
+// OTHER (Fence, Queue SwapChain, Device)
+// ****************************************************************************
+
+// Fence
+interface WebGPUFence {
+    u32 getValue();
+};
+
+// Queue
+interface WebGPUQueue {
+    void submit(sequence<WebGPUCommandBuffer> buffers);
+    void updateFence(WebGPUFence fence, u32 value);
+};
+
+// SwapChain / RenderingContext
+dictionary WebGPUSwapChainDescriptor {
+    WebGPUTextureUsageFlags usage;
+    WebGPUTextureFormatEnum format;
+    u32 width;
+    u32 height;
+};
+
+interface WebGPUSwapChain {
+    void configure(WebGPUSwapChainDescriptor descriptor);
+    WebGPUTexture getNextTexture();
+    void present();
+};
+
+interface WebGPURenderingContext : WebGPUSwapChain {
+};
+
+// Device
+interface WebGPUDevice {
+    WebGPUBuffer createBuffer(WebGPUBufferDescriptor descriptor);
+    WebGPUTexture createTexture(WebGPUTextureDescriptor descriptor);
+    WebGPUSampler createSampler(WebGPUSamplerDescriptor descriptor);
+
+    WebGPUBindGroupLayout createBindGroupLayout(WebGPUBindGroupLayoutDescriptor descriptor);
+    WebGPUPipelineLayout createPipelineLayout(WebGPUPipelineLayoutDescriptor descriptor);
+    WebGPUBindGroup createBindGroup(WebGPUBindGroupDescriptor descriptor);
+
+    WebGPUBlendState createBlendState(WebGPUBlendStateDescriptor descriptor);
+    WebGPUDepthStencilState createDepthStencilState(WebGPUDepthStencilStateDescriptor descriptor);
+    WebGPUInputState createInputState(WebGPUInputStateDescriptor descriptor);
+    WebGPUShaderModule createShaderModule(WebGPUShaderModuleDescriptor descriptor);
+    WebGPUAttachmentState createAttachmentState(WebGPUAttachmentStateDescriptor descriptor);
+    WebGPUComputePipeline createComputePipeline(WebGPUComputePipelineDescriptor descriptor);
+    WebGPURenderPipeline createRenderPipeline(WebGPURenderPipelineDescriptor descriptor);
+
+    WebGPUCommandEncoder createCommandEncoder(WebGPUCommandEncoderDescriptor descriptor);
+    WebGPUFence createFence(WebGPUFenceDescriptor descriptor);
+
+    WebGPUQueue getQueue();
+};
+
+// WebGPU "namespace" used for device creation
+dictionary WebGPUExtensions {
+    bool anisotropicFiltering;
+};
+
+dictionary WebGPUFeatures {
+    bool logicOp;
+};
+
+dictionary WebGPULimits {
+    u32 maxBindGroups;
+};
+
+dictionary WebGPUDeviceDescriptor {
+    WebGPUExtensions extensions;
+    WebGPULimits limits;
+    WebGPUFeatures features;
+    // TODO are other things configurable like queues?
+};
+
+interface WebGPU {
+    static WebGPUExtensions getExtensions();
+    static WebGPUFeatures getFeatures();
+    static WebGPULimits getLimits();
```

Comment:
Exposing device-specific limits encourages fingerprinting. Instead, we should require a certain set of limits are required by any implementation.

Reply:
In terms of fingerprinting, how is `getLimits` any different from most of the queries through `getParameter` (especially those prefixed with `MAX`)? Perhaps we need to revisit the discussion from before because I don't think there was clear consensus on this (performance vs. portability/fingerprinting concerns).